src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs

```diff
@@ -72,12 +72,19 @@ internal abstract partial class HttpProtocol : IHttpResponseControl
         private Stream _requestStreamInternal;
         private Stream _responseStreamInternal;
 
-        public HttpProtocol(HttpConnectionContext context)
+        private readonly bool _useLatin1;
+
+        public HttpProtocol(HttpConnectionContext context, bool isHttp1)
         {
             _context = context;
 
             ServerOptions = ServiceContext.ServerOptions;
-            HttpRequestHeaders = new HttpRequestHeaders(reuseHeaderValues: !ServerOptions.DisableStringReuse);
+            _useLatin1 = isHttp1 && ServerOptions.Latin1RequestHeaders;
```

Comment:
Why just HTTP1? HTTP/2 would be affected the same way.

Reply:
When we talked about it in the meeting with @lodejard, @mjsabby and @timmydo, they said this change was for legacy clients and that requiring UTF-8 for HTTP/2 clients was fine.

Reply:
Did that come up specifically? I don't recall...

Taking a quick look at the HTTP/2 and HPACK specs it doesn't appear that a character set encoding is specified for the header values - all I can see is a reference to them as octets.

I would say the server shouldn't take the transport version into account. Especially because intermediate gateways between client and server could shift the connection between HTTP/1.* and HTTP/2 --- the header values would be the same end-to-end but the http version is only the same hop-by-hop.